liquibase-core/src/main/java/liquibase/changelog/DatabaseChangeLog.java

```diff
@@ -48,6 +46,7 @@ public class DatabaseChangeLog implements Comparable<DatabaseChangeLog>, Conditi
     private String physicalFilePath;
     private String logicalFilePath;
     private String changeLogId;
+    private String includeRunWith;
```

Comment:
I don't think we can be adding an object-level field for includeRunWith. It happens to work with how we tend to call methods on this object, but the scope seems wrong on it. It's not that there's a DatabseChangeLog-wide "includeRunWith", it's really just a scratch storage variable that should be scoped to where it actually applies. Otherwise it's going to be too easy to have bugs when people are calling include() in their own unique ways without knowing to somehow clean this up. Or as we evolve our usage of this class over time it may burn us.

Since we have a new concept of modifyChangeSets we probably need a class to contain that. It will just have runWith as a field on it now, but will allow us to keep adding more flags without breaking API compatibility. Updating the include() function to take the applicable ModifyChangeSets object (but keeping a deprecated vesrsion that doesn't have the parameter for API compatibility) makes it so that function doesn't have to rely on this object-level field.

But there is the question of how to store it on line 399 where we'd be creating it when working on one ParsedNode to where we need to get it on line 404 when being called for a different ParsedNode....

We could add a scratch Map to handleChildNode's signature and on line 399 do a nodeScratch.put("modifyChangeSets", new ModifyChangeSets(...)) and then on line 402 after all the "handleChildNode" call nodeScratch.remove("modifyChangeSets"). That's a bit ugly, but is at least keeping it scoped a bit cleaner? 

Ideally we'd be able to walk up the parent node tree to look for attached/parsed versions of the objects. But that is probably a bigger refactoring than we want right now.

Any other ideas?

Reply:
I have re-worked the implementation.